### PR TITLE
Allow . in bucket names

### PR DESCRIPTION
**Pull request by jneem**
Bucket names in GCS are allowed to contain '.' (under some conditions; see [here](https://cloud.google.com/storage/docs/domain-name-verification)), but `tame-gcs` fails validation for those names, as reported [here](https://github.com/EmbarkStudios/tame-gcs/issues/58).

This PR works around the issue by copying tame-gcs's bucket verification code, and modifying it to allow '.'. It's written with the assumption that tame-gcs will eventually fix the issue and so we can remove this code; if we intended to actually keep this around, it would probably make more sense to integrate `BucketNameError` into `InvalidNameError`.

Also, while testing this I noticed that build fails with `--features=storage`, because of a missing `tower` dependency.

### CODE REVIEW

File: src/storage.rs [hunk 1]
```diff
@@ -37,19 +37,98 @@ async fn collect_body(
 pub enum InvalidNameError {
     /// The given name did not have a valid bucket
     #[error("error validating bucket name {1}")]
-    Bucket(#[source] api::Error, String),
+    Bucket(#[source] BucketNameError, String),
```
Comment (rnarubin) [inline src/storage.rs:40]: To better support adopting tame-gcs's fix (hoping one happens), can we keep the error type the same? tame_gcs::Error has public variants, it should be possible to construct the same `Invalid*` variants

**Review by rnarubin — COMMENTED** [review]
The workaround seems reasonable to me

File: src/storage.rs (hunk 1 above)
Comment (jneem) [reply src/storage.rs]: Good point, I've switched back to `tame-gcs` errors.

File: src/storage.rs [hunk 2]
```diff
@@ -159,7 +211,7 @@ where
 
         let request = objects::Object::download(&oid, None).map_err(ObjectError::InvalidRequest)?;
 
-        let response = self.send_request(empty_body(request)).await?;
+        let response = dbg!(self.send_request(empty_body(request)).await?);
```
Comment (rnarubin) [inline src/storage.rs:214]: a `dbg` has snuck in here

**Review by rnarubin — APPROVED** [review]

**Pull request merged**